homeassistant/components/habitica/calendar.py

```diff
@@ -0,0 +1,217 @@
+"""Calendar platform for Habitica integration."""
+
+from __future__ import annotations
+
+from datetime import date, datetime, timedelta
+from enum import StrEnum
+
+from dateutil.rrule import rrule
+
+from homeassistant.components.calendar import (
+    CalendarEntity,
+    CalendarEntityDescription,
+    CalendarEvent,
+)
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.util import dt as dt_util
+
+from . import HabiticaConfigEntry
+from .coordinator import HabiticaDataUpdateCoordinator
+from .entity import HabiticaBase
+from .types import HabiticaTaskType
+from .util import build_rrule, get_recurrence_rule
+
+
+class HabiticaCalendar(StrEnum):
+    """Habitica calendars."""
+
+    DAILIES = "dailys"
+    TODOS = "todos"
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: HabiticaConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up the calendar platform."""
+    coordinator = config_entry.runtime_data
+
+    async_add_entities(
+        [
+            HabiticaTodosCalendarEntity(coordinator),
+            HabiticaDailiesCalendarEntity(coordinator),
+        ]
+    )
+
+
+class HabiticaCalendarEntity(HabiticaBase, CalendarEntity):
+    """Base Habitica calendar entity."""
+
+    def __init__(
+        self,
+        coordinator: HabiticaDataUpdateCoordinator,
+    ) -> None:
+        """Initialize calendar entity."""
+        super().__init__(coordinator, self.entity_description)
+
+
+class HabiticaTodosCalendarEntity(HabiticaCalendarEntity):
+    """Habitica todos calendar entity."""
+
+    entity_description = CalendarEntityDescription(
+        key=HabiticaCalendar.TODOS,
+        translation_key=HabiticaCalendar.TODOS,
+    )
+
+    def dated_todos(
+        self, start_date: datetime, end_date: datetime | None = None
+    ) -> list[CalendarEvent]:
+        """Get all dated todos."""
+
+        events = []
+        for task in self.coordinator.data.tasks:
+            if (
+                task["type"] == HabiticaTaskType.TODO  # only todos
+                and not task["completed"]  # that are not completed
+                and task.get("date")  # and have a due date
+            ):
+                start = dt_util.start_of_local_day(datetime.fromisoformat(task["date"]))
+                end = start + timedelta(days=1)
+                # return current and upcoming events or events events within the requested range
+                if (not end_date and start_date < end) or (
```

Comment:
The flipped order of the comparison (first `start_date < end` and then `end > start_date` is a little difficult to follow.

How about something like this:
```
if end < start_date:
    # Event ends before date range
    continue
if end_date and start > end_date:
    # Event starts after date range
    continue
```